doc/design/milestone-0.1.1/csv-generation.md

```diff
@@ -0,0 +1,162 @@
+
+```Go
+import "github.com/operator-framework/operator-lifecycle-manager/pkg/api/apis/operators/v1alpha1"
+
+type CSVUpdater interface {
+	Apply(*v1alpha1.ClusterServiceVersion) error
+}
+```
+
+`Apply` will use data from the `CSVUpdater` implementer to operate on `*v1alpha1.ClusterServiceVersion` object fields relevant to that updater. The OLM defines the entire CSV spec [in a Golang struct][olm-csv-spec-code] the SDK can alias to implement `CSVUpdater`s.
+
+Once sub-step two is reached when creating or updating a CSV, `renderCSV` will extract each yaml document discovered, and pass document data into a dispatcher function. The dispatcher selects the correct `CSVUpdater` to call based on the documents' `Kind` yaml object, a manifest type identifier used in all operator manifests.
+
+The following is an example implementation of an [install strategy][olm-csv-install-strat-doc] `CSVUpdater`:
+
+```Go
+import (
+  ...
+  appsv1 "k8s.io/api/apps/v1"
+  rbacv1beta1 "k8s.io/api/rbac/v1beta1"
+  "github.com/operator-framework/operator-lifecycle-manager/pkg/api/apis/operators/v1alpha1"
+  "github.com/operator-framework/operator-lifecycle-manager/pkg/controller/install"
+)
+
+// CSVInstallStrategyUpdate embeds the OLM's install strategy spec.
+type CSVInstallStrategyUpdate struct {
+	*install.StrategyDetailsDeployment
+  
+  // Future utility fields go here.
+}
+
+// getLocalInstallStrategyCache retrieves the local cache singleton and returns the install strategy cache.
+func getLocalInstallStrategyCache() *CSVInstallStrategyUpdate {
+	factory := getLocalCacheFactory()
+	return factory.InstallStrategyCache
+}
+
+// AddDeploymentSpecToCSVInstallStrategyUpdate adds an RBAC Role to the local cache singletons' permissions.
+func AddRoleToCSVInstallStrategyUpdate(yamlDoc []byte) error {
+	localInstallStrategyUpdate := getLocalInstallStrategyCache()
+
+	newRBACRole := new(rbacv1beta1.Role)
+	_ = yaml.Unmarshal(yamlDoc, newRBACRole)
+
+	newPermissions := install.StrategyDeploymentPermissions{
+		ServiceAccountName: newRole.ObjectMeta.Name,
+		Rules:              newRole.Rules,
+	}
+	localInstallStrategyUpdate.Permissions = append(localInstallStrategyUpdate.Permissions, newPermissions)
+
+	return nil
+}
+
+// AddDeploymentSpecToCSVInstallStrategyUpdate adds a Deployment to the local cache singletons' install strategy.
+func AddDeploymentSpecToCSVInstallStrategyUpdate(yamlDoc []byte) error {
+	localInstallStrategyUpdate := getLocalInstallStrategyCache()
+
+	newDeployment := new(appsv1.Deployment)
+	_ = yaml.Unmarshal(yamlDoc, newDeployment)
+  
+	newDeploymentSpec := install.StrategyDeploymentSpec{
+		Name: newDeployment.ObjectMeta.Name,
+		Spec: newDeployment.Spec,
+	}
+	localInstallStrategyUpdate.DeploymentSpecs = append(localInstallStrategyUpdate.DeploymentSpecs, newDeploymentSpec)
```

Comment:
I get the idea here but just wanted to point out that I don't think we'll be appending the deployment to the install strategy of the CSV in case we make an update to `deploy/operator.yaml`
We'll be overwriting that section since there's usually just 1 deployment per operator(which is the case for the SDK).
e.g https://github.com/operator-framework/operator-lifecycle-manager/blob/8799f39ef342dc1ff7430eba7a88c1c3c70cbdcc/deploy/upstream/manifests/0.7.0/08-rh-operators.configmap.yaml#L7997-L8044

Multiple deployments are for when you're bundling together 3 operators in 1 csv(like we did for the etcd-operator and related operators backup-operator and restore-operator).

Reply:
In this example, `localInstallStrategyUpdate` is a local cache initiated on each `operator-sdk build`. It does not contain any deployments (or permissions as per your other comment) from the current `csv.yaml`; it discovers all `Kind: Deployment` files and appends them to `localInstallStrategyUpdate.DeploymentSpecs` to cover the multiple-deployment case you mention. None of this is obvious from this code example. I will add a detailed description of how the `localInstallStrategyCache` factory works.